src/blog/speeding-up-v8-heap-snapshots.md

```diff
@@ -0,0 +1,161 @@
+ - tools
+---
+*This blog post has been authored by José Dapena Paz (Igalia), with contributions from Jason Williams (Bloomberg), Ashley Claymore (Bloomberg), and Rob Palmer (Bloomberg).*
+
+In this post about V8 heap snapshots, I will talk about some performance problems found by Bloomberg engineers, and how we fixed them to make JavaScript memory analysis faster than ever.
+
+## The Problem
+
+Ashley Claymore was working on diagnosing a memory leak in a JavaScript application. It was failing with *Out-Of-Memory* errors. Despite the process having access to plenty of system memory, V8 places a hard limit on the amount of memory dedicated to the garbage-collected heap from which all JavaScript objects are allocated. This V8 heap limit (~1400MB) was being hit.
+
+The standard way to debug a routine memory leak scenario like this is to capture a heap snapshot and then inspect the various summaries and object attributes using DevTools "Memory" tab to find out what is consuming the most memory.  In DevTools, you click the round button marked _"Take heap snapshot"_ to perform the capture. For Node.js applications, you can [trigger the snapshot](https://nodejs.org/en/docs/guides/diagnostics/memory/using-heap-snapshot) programmatically using this API:
+
+```javascript
+require('v8').writeHeapSnapshot();
+```
+
+The desire was to capture several snapshots at different points in the application's life, so that DevTools Memory viewer could be used to show the difference between each.  The problem was that capturing a single full-size (500MB) snapshot was taking **over 30 minutes** alone!
+
+It was this slowness in the memory analysis workflow that we needed to solve.
+
+## Narrowing down the Problem
+
+Jason Williams started investigating the issue using some V8 parameters. As described in the previous post, V8 has some nice command line parameters that can help with that. These options were used to curate the heap snapshots , simplify the reproduction, and improve observability:
+
+`--max-old-space-size=100`
+: This limits the heap to 100 megabytes and helps to reproduce the issue much faster.
+
+`--heapsnapshot-near-heap-limit=10`
+: This tells V8 to produce a maximum of 10 snapshots, each time it comes close to running out of memory.  This prevents thrashing where the memory-starved program spends a long time producing more snapshots than needed.
+
+`--enable-etw-stack-walking`
+: This allows tools such as ETW, WPA & xperf to see the JS stack which has been called in V8. (Node v20+)
+
+`--interpreted-frames-native-stack`
+: This flag is used in combination with tools like ETW, WPA & xperf to see the native stack when profiling. (Node v20+).
+
+After each snapshot, V8 tries to force garbage collection to reduce memory usage and avoid hitting the limit. In the test case, the memory usage increases, but, after several iterations, garbage collection ultimately could not free up enough space and so the application terminated with an *Out-Of-Memory* error.
```

Comment:
This is interesting: the near limit heap snapshot callback is invoked in several scenarios. Most of them are exactly happening on different garbage collection stages (the exception is when it has an ineffective MarkCompact, that is usually pretty bad.

So, I think we need to rewrite this paragraph to be more accurate. Something like...

> When V8 is close to the heap limit, it forces a garbage collection to reduce the memory usage and avoid hitting the limit. With the provided Node.js `--heapsnapshot-near-heap-limit` flag it will also dump a new heap snapshot. In the test case, the memory usage increases, but, after several iterations, garbage collection ultimately can not free up enough space and so the application is terminated with an *Out-Of-Memory* error.

Reply:
LGTM, except

> With the provided Node.js --heapsnapshot-near-heap-limit flag it will also dump a new heap snapshot. 

could be tweaked to

> It also notifies the embedder that the heap is about to reach the memory limit. The `--heapsnapshot-near-heap-limit` flag in Node.js dumps a new heap snapshot upon notification.

Otherwise it reads as if V8 is the one initiating the heap snapshot generation.

(Also I think you probably meant `the memory usage decreases, but..` :)).

Reply:
Using all these comments. Thanks!